File: python/binary/binary_search.py
'''
в переменной low и high храняться границы 
той части списка, в которой выполняется поиск
'''
def binary_search(list, item):
    low = 0
    high = len(list) - 1

    while low <= high:  #пока эта часть не сократиться до одного элемента
        #проверяем среднее значение
        mid = (low + high) // 2
        guess = list[mid]
        
        #значение найдено
        if guess == item:
            return mid
        
        #много
        if guess > item:
            high = mid - 1
        #мало    
        else:
            low = mid + 1
    return None

my_list = [1, 3, 5, 6, 7]
#получаем номер искомого элемента в спике (в данном случае 4)
print(binary_search(my_list,7)) 

File: kotlin/binary/binarySearch.kt
val a = listOf(1, 2, 3, 4, 5, 6, 7)

fun main() {
    binarySearch(a, 6)
    resultTest(
        list = a,
        testValue = 5
    )
}

fun resultTest(list: List<Int>, item: Int = 6, testValue: Int = 6) {
    if (binarySearch(list, item) == testValue) println("passed")
    else println("false")
}

fun binarySearch(list: List<Int>, item: Int): Int? {
    var low = 0
    //возвращаем индекс последнего элемента
    var high = list.size - 1
    while (low <= high) {
        val mid = (low + high) / 2
        val guess = list[mid]

        if (guess == item) {
            return mid
        }
        if (guess > item) {
            high = mid - 1
        } else {
            low = mid + 1
        }
    }
    return null
}
